### PR TITLE
Null date are now after any date.

### PR DESCRIPTION
**Pull request by theofidry**
If we do a DQL query like `dummy.dummyDate <= :dateValue`, if `dummyDate` is null it will match this criteria. However a date field at null is often used to define an undetermined future date, i.e. a NULL date is expected to be higher than any date.

However I'm still concerned that some may which to keep a null date value as the lower value possible for a date. @dunglas what do you think? Should we add another parameter to configure this?

Also removed some duplication in code.

### CODE REVIEW

**Comment by ogizanagi** [comment]
IMO, null dates should simply be excluded as default behavior. But as their are many legit use cases, I think it should be configurable at the filter level:

```php
# app/config/services.yml

services:
    ressource.date_filter:
        parent:    "api.doctrine.orm.date_filter"
        arguments: [ ["dateProperty"], NEW_OPTION ] #Here

    resource.offer:
        parent:    "api.resource"
        arguments: [ "AppBundle\Entity\Offer" ]
        calls:
            -      method:    "addFilter"
                   arguments: [ "@resource.offer.date_filter" ]
        tags:      [ { name: "api.resource" } ]
```

What do you think ?

File: Doctrine/Orm/DateFilter.php [hunk 1]
```diff
@@ -35,7 +38,7 @@ class DateFilter extends AbstractFilter
     public function __construct(ManagerRegistry $managerRegistry, array $properties = null)
     {
         $this->managerRegistry = $managerRegistry;
-        $this->properties = $properties;
+        $this->properties      = $properties;
```
Comment (dunglas) [inline Doctrine/Orm/DateFilter.php]: It's not significant but in our CS we don't align equals.

**Comment by theofidry** [comment]
Agreed. Either passing it as an argument but I would prefer adding an init method:

```yaml
# app/config/services.yml

services:
    ressource.date_filter:
        parent:    "api.doctrine.orm.date_filter"
        arguments: [ ["dateProperty"], NEW_OPTION ] #Here
        calls:
            - [ "initOrder", [ <"first"|"last"> ] ]
```

Just the question of the default order. Any though on it?

File: Doctrine/Orm/DateFilter.php (hunk 1 above)
Comment (theofidry) [inline Doctrine/Orm/DateFilter.php]: Ok, it's not significant but consistency, even in coding style is always a must :D

**Comment by dunglas** [comment]
I'm :+1: for @ogizanagi's idea too.

In the current implementation their is something weird: null values are excluded for `before` but included for `after`. It's probably good for your use case but this is not generic enough. I think this must be configurable, something like that:

```yaml
# app/config/services.yml

services:
    ressource.date_filter:
        parent:    "api.doctrine.orm.date_filter"
        arguments: [ ["dateProperty"], 0, 2 ] # Class constants
```

```php
// ... 
class DateFilter
{
    const NULL_EXCLUDED = 0;
    const NULL_FIRST = 1;
    const NULL_LAST = 2;
  
    public function __construct(array $properties, $beforeNullValues, $afterValues)
    {
        // ...
    }
}
```

**Comment by dunglas** [comment]
Btw, I prefer constructor parameter than init method when possible. It's OK here IMO.

**Comment by ogizanagi** [comment]
We could imagine it'll became something quite common to have different kind of options in order to configure filters behavior. So why not making the second argument an array of options ? It has the benefit of being more explicit and readable in services declarations:

```php
# app/config/services.yml
services:
    ressource.date_filter:
        parent:    "api.doctrine.orm.date_filter"
        arguments: 
            - ["dateProperty"]
            - 
              before_null_values: 0
              after_null_values: 2
```
```php
# DateFilter.php
class DateFilter
{
    const NULL_EXCLUDED = 0;
    const NULL_FIRST = 1;
    const NULL_LAST = 2;

    $defaultOptions = array(
        'before_null_value' => self::NULL_EXCLUDED,
        'after_null_value' => self::NULL_LAST,
    );

    public function __construct(Man..., array $properties, array $options = array())
    {
        $options = array_replace($this->defaultOptions, $options);
        $beforeNullValues = $options['before_null_values'];
        $afterNullValues = $options['after_null_values'];
        // ...
    }
}
```

**Comment by theofidry** [comment]
ping @dunglas: should be good now.

File: Doctrine/Orm/DateFilter.php [hunk 2]
```diff
@@ -23,19 +23,35 @@
...
+     * @var array List of properties by witch the collection can be ordered.
+     */
+    protected $properties;
+
     /**
-     * @var array List of properties by witch the collection can or cannot be ordered.
+     * @var int Specify the NULL position when sorting dates
      */
-    private $properties;
+    protected $nullOption = self::NULL_FIRST;
```
Comment (dunglas) [inline Doctrine/Orm/DateFilter.php]: Should be private by default.

File: Doctrine/Orm/DateFilter.php [hunk 3]
```diff
@@ -47,36 +63,71 @@ public function __construct(ManagerRegistry $managerRegistry, array $properties
...
+                                    sprintf('o.%s <= :%s', $filter, $parameter),
+                                    $queryBuilder->expr()->isNotNull(sprintf('o.%s', $filter))
+                                );
+                            } elseif (self::PARAMETER_AFTER === $period) {
+                                $queryBuilder->andWhere(sprintf('o.%s >= :%s', $filter, $parameter));
+                            }
+                            break;
+
+                        case self::NULL_FIRST:
+                            if (self::PARAMETER_BEFORE === $period) {
```
Comment (dunglas) [inline Doctrine/Orm/DateFilter.php]: A switch seems more appropriated than an if / else block in that case.

File: Tests/Doctrine/Orm/DateFilterTest.php [hunk 4]
```diff
@@ -93,36 +94,129 @@ public function filterProvider()
     {
         return [
             // Properties enabled with valid values
+            // Test after
             [
                 [
                     'properties' => null,
+                    'nullOption' => DateFilter::NULL_EXCLUDED
```
Comment (dunglas) [inline Tests/Doctrine/Orm/DateFilterTest.php]: missing comma at the end of a line (but run PHP CS Fixer, it will fix it for you).

**Comment by dunglas** [comment]
Some minor issues but looks great @theofidry

File: Doctrine/Orm/DateFilter.php (hunk 3 above)
Comment (theofidry) [inline Doctrine/Orm/DateFilter.php]: You mean to replace the if/else block within each cases by a switch?

Comment (dunglas) [inline Doctrine/Orm/DateFilter.php]: yes

**Comment by theofidry** [comment]
It should be good now. Tell me if it is so that I rebase once before merging.

**Comment by dunglas** [comment]
Ok for me. Please run CS fixer then squash your commits and I'll merge it.

Thank you Théo.

**Comment by theofidry** [comment]
Done :)

**Comment by dunglas** [comment]
Hum we just missed one thing: can you update the doc please @theofidry?

**Comment by theofidry** [comment]
Ah indeed, right away

**Pull request merged**